zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,191 @@
+        let r_i = Nonce(Scalar::random(&mut rng));
+        // R_i = generator * r_i
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i.0);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript,
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+                nonce_precommitments,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Siglet), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            // TBD: should we use ct_eq?
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
+            }
+        }
+
+        // Make R = sum_i(R_i). nonce_commitments = R_i from all the parties.
+        let R: RistrettoPoint = nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        // Make c = H(X_agg, R, m)
+        // The message should already have been fed into the transcript.
+        let c = {
+            self.transcript
+                .commit_point(b"X_agg", &self.multikey.aggregated_key().0);
+            self.transcript.commit_point(b"R", &R.compress());
+            self.transcript.challenge_scalar(b"c")
+        };
+
+        // Make a_i = H(L, X_i)
+        let X_i = VerificationKey((self.x_i.0 * RISTRETTO_BASEPOINT_POINT).compress());
+        let a_i = self.multikey.factor_for_key(&X_i);
+
+        // Generate siglet: s_i = r_i + c * a_i * x_i
+        let s_i = self.r_i.0 + c * a_i * self.x_i.0;
+
+        // Store received nonce commitments in next state
+        Ok((
+            PartyAwaitingSiglets {
+                multikey: self.multikey,
+                nonce_commitments,
+                c,
+            },
+            Siglet(s_i),
+        ))
+    }
+}
+
+impl PartyAwaitingSiglets {
+    pub fn receive_siglets(self, siglets: Vec<Siglet>) -> Signature {
```

Comment:
maybe `receive_trusted_siglets` name for method that does not verify siglets, and `receive_siglets` for method that does (cf. range proof API: https://github.com/dalek-cryptography/bulletproofs/blob/main/src/range_proof/dealer.rs#L328)?

Also: would it be more clear or less if we rename "siglets" into "shares" like with rangeproof API?

Reply:
siglets technically aren't "signature shares", because they don't contain the `R_i` components... so it's not quite the same as the rangeproof API. I don't have a strong opinion on calling them "siglets" or "shares".

I renamed the siglet receiving functions, that makes sense!

Reply:
Oh, now that I have gotten rid of the `Siglet` struct, there is less context on what "siglet" means. So, calling the functions `...shares` instead of `...siglets` makes a lot more sense now. Updated the naming!